network/iptables.go

```diff
@@ -162,11 +169,23 @@ func setupIPTables(ipt IPTables, rules []IPTablesRule) error {
 	return nil
 }
 
-func teardownIPTables(ipt IPTables, rules []IPTablesRule) {
+func teardownIPTables(ipt IPTables, rules []IPTablesRule) error {
 	for _, rule := range rules {
 		log.Info("Deleting iptables rule: ", strings.Join(rule.rulespec, " "))
-		// We ignore errors here because if there's an error it's almost certainly because the rule
-		// doesn't exist, which is fine (we don't need to delete rules that don't exist)
-		ipt.Delete(rule.table, rule.chain, rule.rulespec...)
+		err := ipt.Delete(rule.table, rule.chain, rule.rulespec...)
+		if err != nil {
+			e := err.(IPTablesError)
+			// If this error is because the rule is already deleted, the message from iptables will be
+			// "Bad rule (does a matching rule exist in that chain?)". These are safe to ignore.
+			// However other errors (like EAGAIN caused by other things not respecting the xtables.lock)
+			// should halt the ensure process.  Otherwise rules can get out of order when a rule we think
+			// is deleted is actually still in the chain.
+			// This will leave the rules incomplete until the next successful reconciliation loop.
+			if !e.IsNotExist() {
+				return err
```

Comment:
Should we not aggregate the errors and return them after running through the loop? We do not want a bad rule to block others from being removed. That can result in `all` rules staying stuck because of one broken one.

Reply:
This presumes that the error deleting the rule was transient and that if we bail out here then it is likely to succeed on the next iteration of the reconciliation loop.  Deleting other rules after a failure is what complicates things when trying to reinsert the rules in the right order. If there’s a rule that consistently fails to delete I don’t know how much it matters that it blocks the others since it would be difficult to guarantee correct behavior in that state anyway.